src/Webinex.Calendar/Filters/EventFiltersProvider.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System.Linq.Expressions;
+using Webinex.Calendar.Common;
+using Webinex.Calendar.DataAccess;
+using Webinex.Calendar.Events;
+
+namespace Webinex.Calendar.Filters;
+
+public record EventFiltersProvider<TData>(
+    DateTimeOffset From,
+    DateTimeOffset To,
+    Expression<Func<TData, bool>>? DataFilter,
+    string TimeZone,
+    bool OneTime,
+    bool DayOfMonth,
+    bool DayOfWeek,
+    bool Interval,
+    bool State,
+    bool Data,
+    bool Precise)
+    where TData : class, ICloneable
+{
+    public bool Data { get; set; } = Data;
+    public bool Precise { get; set; } = Precise;
+
+    public Expression<Func<EventRow<TData>, bool>> Create()
+    {
+        Expression<Func<EventRow<TData>, bool>> @base = x =>
+            x.Effective.Start < To.TotalMinutesSince1990() &&
+            (x.Effective.End > From.TotalMinutesSince1990() || x.Effective.End == null);
+
+        if (Data && DataFilter != null)
+            @base = Expressions.And(@base, Expressions.Child<EventRow<TData>, TData>(x => x.Data, DataFilter));
+
+        return Expressions.Or(
+            Expressions.And(@base,
+                Expressions.Or(OneTimeEventFilter() ?? (_ => false), CreateRecurrentEventFilter() ?? (_ => false))),
+            CreateRecurrentEventStateFilter() ?? (_ => false));
+    }
+
+    private Expression<Func<EventRow<TData>, bool>>? OneTimeEventFilter()
+    {
+        if (!OneTime)
+            return null;
+
+        return x => x.Effective.End > From.TotalMinutesSince1990() && x.Type == EventType.OneTimeEvent;
```

Comment:
actually, `x.Effective.End > From.TotalMinutesSince1990()` is in `@base`

Reply:
removed